[Aerie-1686]  Setup JS Command Expansion Server

goetzrrGit opened the pull request:
* **Tickets addressed:** AERIE-1686
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

Instead of having another GitHub repo for the command expansion service, it was decided to keep all the Aerie services under one project. I have added a command-expansion-server directory containing a basic node server we can build upon.

- [x] Gradle build processes
- [x] Docker compose and env variables to connect to db
- [x] PR building
- [x] Publishing build products (docker image to ghcr)
- [x] Identify a testing harness (we can chat about this more if you like)

## Verification

- `docker system prune --all --volumes  <-- Nuke everything `
- `./gradlew build; docker compose -f docker-compose.yml build; docker compose -f docker-compose.yml up`
- Verify you have a new docker container `aerie_commanding` running 
- Login to the new container and see that the node server is running
- Enter localhost:3000 on your browser to connect with the server

## Documentation
Added Command Expansion Service -  https://github.com/NASA-AMMOS/aerie/wiki/Product-Guide

## Future work
[AERIE-1687](https://jira.jpl.nasa.gov/browse/AERIE-1687)
Setup database schema intialization for command expansion server

goetzrrGit committed:
Added the `Command Expansion Service` project

Instead of having another GitHub repo for the command expansion service, it was decided to keep all the Aerie services under one project. I have added a command-expansion service directory containing a basic node server we and build upon.

I have also hooked up gradle to build the javascript node project using `./gradlew build`.

[AERIE-1686]

goetzrrGit committed:
Updated GitHub actions to build the `command expansion service`
when opening a PR

This is slightly different from the java services because there is no `Test and Assemble` gradle task for the node project.

[AERIE-1686]

goetzrrGit committed:
Created a docker image for `Command Expansion Service`

These environment variables have already been hooked up in the node project under `env.ts`. The docker compose will now set the environment values so the command expansion server can use those instead of the default defined in the `env.ts`

[AERIE-1686]

dyst5422 commented on command-expansion/tsconfig.json:
```diff
@@ -0,0 +1,101 @@
+    // "useDefineForClassFields": true,                  /* Emit ECMAScript-standard-compliant class fields. */
+
+    /* Modules */
+    "module": "commonjs",                                /* Specify what module code is generated. */
```
We should target ESNext so we run it as a module. Lets tag up quick about this at some point and talk it through.

dyst5422 replied:
Part of this will be renaming files to `.mts` so they transpile to `.mjs` and node recognizes them as module-type files

camargo replied:
For completeness, another option here is to use the `"type": "module"` flag in your `package.json` file:
https://nodejs.org/docs/latest-v16.x/api/packages.html#type
If you want to drop back to CommonJS you can use the `.cjs` extension.

dyst5422 replied:
The support in the package.json file is sporadic among tooling.

camargo replied:
What tooling? This feature is very widely used in big open source JS projects

dyst5422 replied:
Jest is the one that jumps out at me (although it's been a year or so since i've tried doing that).

dyst5422 commented on command-expansion/tsconfig.json:
```diff
@@ -0,0 +1,101 @@
+    // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
+
+    /* Language and Environment */
+    "target": "es2016",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
```
We should have this be ESNext to track any JS api updates - since we're pulling the latest node version in the docker image

camargo replied:
Note we are using the LTS Node version for all our Docker base images (v16.4.0).

goetzrrGit replied:
@dyst5422 @camargo Will `ESNext` use the LTS version or the latest version which could be something higher in the future?

dyst5422 replied:
It tracks the latest advancements to the language (that have been stabilized)

`Additional APIs available in ESNext - This changes as the JavaScript specification evolves`

camargo replied:
So I believe `ESNext` currently points to `ES2022+`: https://www.typescriptlang.org/tsconfig#lib
According to this site https://node.green/#ES2022 Node LTS for `ES2022` is ~70% feature complete.
Where `ES2021` is 100% feature complete: https://node.green/#ES2021 so I'd go with that

dyst5422 replied:
Sounds good

dyst5422 commented on command-expansion/tsconfig.json:
```diff
@@ -0,0 +1,101 @@
+
+    /* Language and Environment */
+    "target": "es2016",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
+    // "lib": [],                                        /* Specify a set of bundled library declaration files that describe the target runtime environment. */
```
`"lib": ["node"],`

goetzrrGit replied:
@dyst5422 there is no `"lib": ["node"]` available. I can set it to `ES2021` though. Would that work? Since the `target` and `module` are `ES2021`

dyst5422 replied:
Right! sorry, I am tired today and not thinking straight. Yes thats fine. I think we just want the latest stable specification for each of these.

dyst5422 commented on command-expansion/tsconfig.json:
```diff
@@ -0,0 +1,101 @@
+    /* Modules */
+    "module": "commonjs",                                /* Specify what module code is generated. */
+     "rootDir": "./src",                                  /* Specify the root folder within your source files. */
+    // "moduleResolution": "node",                       /* Specify how TypeScript looks up a file from a given module specifier. */
```
Uncomment this

dyst5422 commented on command-expansion/tsconfig.json:
```diff
@@ -0,0 +1,101 @@
+    // "typeRoots": [],                                  /* Specify multiple folders that act like `./node_modules/@types`. */
+    // "types": [],                                      /* Specify type package names to be included without being referenced in a source file. */
+    // "allowUmdGlobalAccess": true,                     /* Allow accessing UMD globals from modules. */
+    // "resolveJsonModule": true,                        /* Enable importing .json files */
```
Uncomment this for easy building in of JSON configs

dyst5422 replied:
Sorry, I don't know why its tagging ALL these lines, I meant just the last one with `resolveJsonModule`

The others we can leave out

mattdailis commented on settings.gradle:
```diff
@@ -14,6 +14,7 @@ include ':parsing-utilities'
 // Services for deployment within the Aerie infrastructure
 include ':merlin-server'
 include ':scheduler-server'
+include ':command-expansion'
```
Would it make sense to follow this naming convention and call it `command-expansion-server`?

goetzrrGit replied:
Sure no objections to the name

dyst5422 commented:
Regarding a test runner, Jest is definitely the way to go nowadays. There are a couple steps to setting it up with TypeScript

Additional dev-dependencies for that - `jest`, `jest-circus`, `jest-html-reporters`, `ts-jest`

Probably also want to include `ts-doc` for auto documentation generation

mattdailis commented on command-expansion/build.gradle:
```diff
@@ -0,0 +1,25 @@
+task build(type: NpxTask) {
+  dependsOn npmInstall
+
+  command = 'npx'
```
I'm curious about the pros/cons of putting together the command here, vs using the npm run tasks directly.
(see [docs](https://github.com/node-gradle/gradle-node-plugin/blob/master/docs/usage.md#executing-npm-tasks))

For example, depending on the `npm_build` task would be equivalent to running `npm run build`, which is defined to be `npx tsc` in package.json.

One possible advantage of your way is that specifying the input files could prevent unnecessary re-builds? But I'm just guessing here.

goetzrrGit replied:
I wanted to make things simpler and consistent by using the same command `./gradlew build` which will build all the other Aerie services. You can `./gradlew npm_build` too, but I felt that was one extra thing to remember so I defined a `build` task instead.

mattdailis replied:
Good point :+1: I like the consistency as well.

I'm noticing that the first line is `dependsOn npmInstall`. Would it still achieve what we're trying to do if we added another line `dependsOn npm_build`?

I'm not super familiar with gradle, so I'm not sure this is the "right way" to call one task from another task. I'm curious to hear your thoughts.

goetzrrGit replied:
Thanks for pointing this out. I reworked the task into npm type which is a lot cleaner

camargo commented on command-expansion/src/app.js:
```diff
@@ -0,0 +1,11 @@
+"use strict";
```
Since this `.js` file is built (generated by `tsc`) we can/should remove it

goetzrrGit replied:
Actually, this file shouldn't be here so I removed it.

goetzrrGit committed:
Updated GitHub actions to build a docker image for `Command Expansion Services`

This should build the command expansion service and push the docker image to GitHub packages when a PR is merged into develop

[AERIE-1686]

pcrosemurgy commented on command-expansion-server/build.gradle:
```diff
@@ -0,0 +1,19 @@
+}
+
+task clean(type: Delete) {
+  delete '.gradle', 'build','node_modules', 'package-lock.json'
```
I think you can safely remove `'.gradle'` from this list here. `.gradle/` should only contain cached Gradle dependencies, and in this Gradle subproject the only cached dep. should be the `node-gradle` plugin.

pcrosemurgy replied:
I also think `'build'` can be removed since this task (`clean`) is not overriding the default Gradle `clean` task (which removes `build/`). I'd double check me on this though.

goetzrrGit replied:
Thanks for the suggestion. I should make a task called `nuke` which will remove everything. :P

goetzrrGit replied:
I just verified and `'build'` is needed. `.gradle clean` doesn't remove the `build/` directory in the javascript project.

pcrosemurgy replied:
Thanks for checking! :)

camargo commented on command-expansion-server/src/app.mts:
```diff
@@ -0,0 +1,18 @@
+import express, { Application, Request, Response } from "express";
```
We lost syntax highlighting here with the `.mts` extension. 
Can we consider the other option of using `"type": "module"` in `package.json`?

mattdailis replied:
I don't have an opinion on `.mts` vs "type": "module", since all this is very new to me, and I'm still learning about what javascript modules even are.

However, the syntax highlighting doesn't seem to be an issue on my end - VSCode was able to recognize `.mts` as typescript out of the box, and IntelliJ let me add `*.mts` to the list of patterns associated with Typescript under `Preferences/Editor/File Types`.

dyst5422 replied:
We had a discussion on slack (@camargo @goetzrrGit @dyst5422) and resolved all concerns with type: module, so I think we'll go that direction to preserve highlighting in GH

goetzrrGit replied:
@dyst5422 it should be fixed now.

pcrosemurgy commented:
Making a note to myself: on #51 https://github.com/NASA-AMMOS/aerie/pull/51/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7R16 we'll need to add a line here to detect `processResources` for the `command-expansion-server` subproject. Kind of confusing but basically `build.gradle` has no knowledge of whether each subproject has an actual `processResources` task unless a narrowing `plugins.withId()` context is used.

Ryan, I'll work with you to make sure we get this right :)

pcrosemurgy reviewed (approved):
My JS is not too strong but the DB connection setup/logic seems to make sense.

dyst5422 commented on command-expansion-server/src/packages/db/db.ts:
```diff
@@ -0,0 +1,40 @@
+    return DbExpansion.pool;
+  }
+
+  static async init(): Promise<void> {
```
I don't see any async in this function right now. I assume we have this as an async method assuming that there is potential for it to be async in the future.

goetzrrGit replied:
Right now we are not doing anything async so I will remove it. If we do later, I can add it back in.

dyst5422 commented on docker-compose.yml:
```diff
@@ -64,6 +64,24 @@ services:
+      MERLIN_GRAPHQL_URL: http://hasura:8080/v1/graphql
+      COMMANDING_SERVER_PORT : 3000
+      COMMANDING_DB: "aerie_expansion"
+      COMMANDING_DB_PASSWORD: "aerie"
```
Is there security vulnerability with this password being on GH?

goetzrrGit replied:
If you look at the docker-compose we have embedded multiple usernames and passwords for our other services. It is a valid point but maybe not a concern as of now. What do you think @patkenneally

patkenneally replied:
This is on our radar. We should endeavor to remove these in favor of say [Docker Compose Secrets](https://docs.docker.com/engine/swarm/secrets/#build-support-for-docker-secrets-into-your-images) . Additionally, our documentation regarding such deployment considerations is absent. We should provide users with guidance of typical configuration and security considerations. We'll take up this work as a separate ticket.

kroffo reviewed (commented):
I'm going to just comment to say I have looked at this, but I don't have enough expertise in this area to feel like my approval would carry any weight at this time. Since you already have the required number of approvals, this should not be an issue.

pcrosemurgy committed:
Correctly register `command-expansion-server` with SQL distribution

goetzrrGit committed:
Disable testing for until we hook up testing with the command expansion service

[AERIE-1687]
